### PR TITLE
Fix lint errors

### PR DESCRIPTION
**Pull request by finomen**
* Fix segfault when run from CLion
* Fix false positive on template dependent type conversions

### CODE REVIEW

File: clang-tools-extra/clang-tidy/daedalean/TypeConversionsCheck.cpp [hunk 1]
```diff
@@ -207,7 +207,9 @@ void TypeConversionsCheck::handleImplicitCast(clang::ASTContext *context,
                                               QualType sourceType,
                                               QualType destType,
                                               SourceLocation location) {
-  if (destType.getCanonicalType() == sourceType.getCanonicalType()) {
+  sourceType = sourceType.getCanonicalType();
+  destType = destType.getCanonicalType();
+  if (destType == sourceType) {
```
Comment (Javier-varez) [inline clang-tools-extra/clang-tidy/daedalean/TypeConversionsCheck.cpp]: I wonder why they have this  distinction between canonical and non-canonical types...

File: clang-tools-extra/clang-tidy/daedalean/IncludeOrderCheck.cpp [hunk 2]
```diff
@@ -161,11 +161,15 @@ void IncludeOrderPPCallbacks::InclusionDirective(
...
+    // This is not regular file
+    return;
+  }
+
   const std::string CurrentDir =
-      SM.getFileManager()
-          .getCanonicalName(
-              SM.getFileEntryForID(SM.getFileID(HashLoc))->getDir())
-          .str();
+      SM.getFileManager().getCanonicalName(FileEntry->getDir()).str();
```
Comment (Javier-varez) [inline clang-tools-extra/clang-tidy/daedalean/IncludeOrderCheck.cpp:172]: This seems unrelated, could you explain what the problem is here?

Comment (finomen) [reply clang-tools-extra/clang-tidy/daedalean/IncludeOrderCheck.cpp:172]: It crashes on preprocessed source from clion because there are few fake includes that has no files and passed via other methods

File: clang-tools-extra/clang-tidy/daedalean/TypeConversionsCheck.cpp (hunk 1 above)
Comment (finomen) [reply clang-tools-extra/clang-tidy/daedalean/TypeConversionsCheck.cpp]: For example some checkers not complaining on template arguments or apply different rules

**Comment by Javier-varez** [comment]
Could you fix the tests? and if possible add test cases for the code snippet that triggered the issue?

File: clang-tools-extra/clang-tidy/daedalean/TypeConversionsCheck.cpp (hunk 1 above)
Comment (Javier-varez) [reply clang-tools-extra/clang-tidy/daedalean/TypeConversionsCheck.cpp]: In my opinion that is a characteristic of the ast node, not part of the type of the node. Although I guess I just don't have enough context about this.

File: clang-tools-extra/clang-tidy/daedalean/TypeConversionsCheck.cpp [hunk 3]
```diff
@@ -207,74 +207,74 @@ void TypeConversionsCheck::handleImplicitCast(clang::ASTContext *context,
...
     // Removing qualifiers by value (not reference or pointer types) is ok
     return;
   }
 
-  if (sourceType->canDecayToPointerType()) {
-    const QualType decayedType = context->getDecayedType(sourceType);
-    if (decayedType.getCanonicalType() == destType.getCanonicalType()) {
+  if (canonicalSourceType->canDecayToPointerType()) {
+    const QualType decayedType = context->getDecayedType(canonicalSourceType);
+    if (decayedType.getCanonicalType() == canonicalDestType.getCanonicalType()) {
```
Comment (Javier-varez) [inline clang-tools-extra/clang-tidy/daedalean/TypeConversionsCheck.cpp]: this is redundant

File: clang-tools-extra/clang-tidy/daedalean/TypeConversionsCheck.cpp [hunk 4]
```diff
@@ -207,74 +207,74 @@ void TypeConversionsCheck::handleImplicitCast(clang::ASTContext *context,
                                               QualType sourceType,
                                               QualType destType,
                                               SourceLocation location) {
-  sourceType = sourceType.getCanonicalType();
-  destType = destType.getCanonicalType();
-  if (destType == sourceType) {
+  const QualType canonicalSourceType = sourceType.getCanonicalType();
```
Comment (Javier-varez) [inline clang-tools-extra/clang-tidy/daedalean/TypeConversionsCheck.cpp:210]: Should we handle static cast in the same manner?

Comment (finomen) [reply clang-tools-extra/clang-tidy/daedalean/TypeConversionsCheck.cpp:210]: No idea, I could not make test of small size even for this

Comment (Javier-varez) [reply clang-tools-extra/clang-tidy/daedalean/TypeConversionsCheck.cpp:210]: Ok, let's not do it for now and we can handle it later if needed.

**Review by Javier-varez — APPROVED** [review]

**Comment by Javier-varez** [comment]
Please, wait for the latest pre-submit checks to finish before merging this.
